enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,264 @@
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+    // will be looking for during detection
+    DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+    Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    MinDeviceCount int `json:"minDeviceCount"`
+    // +optional
+    // Maximum number of Devices that needs to be detected per node.
+    MaxDeviceCount int `json:"maxDeviceCount"`
+    // VolumeMode determines whether the PV created is Block or Filesystem
+    // + optional
+    VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+    // FSType type to be used when volumeMode is Filesystem
+    // +optional
+    FSType string `json:"fsType,omitempty"`
+    // DeviceInclusionSpec are the filtration rules for including a device in the device discovery
+    // +optional
+    DeviceInclusionSpec *DeviceInclusionSpec `json:"deviceInclusionSpec"`
+}
+
+// DeviceMechanicalProperty holds the device's mechanical spec. It can be rotational, nonRotational or
+// RotationalAndNonRotational
+type DeviceMechanicalProperty string
+
+const (
+    // The mechanical properties of the devices
+    // Rotational refers to magnetic disks
+    Rotational DeviceMechanicalProperty = "rotational"
+    // NonRotational refers to ssds
+    NonRotational DeviceMechanicalProperty = "nonRotational"
+    // RotationalAndNonRotational refers to both magnetic and ssds
+    RotationalAndNonRotational DeviceMechanicalProperty = "rotationalAndNonRotational"
```

Comment:
This value represents when you don't want to filter out based on the rotational or non-rotational values, right? Seems like the empty string would more simply represent not to filter on this field.

Reply:
hmm makes sense.

Reply:
done